### PR TITLE
Consistent window view when exits Bufstop window

### PR DESCRIPTION
**Pull request by bruhtus**
I have noticed that sometimes the current line position got changed when exit Bufstop window. Sometimes the current line position at the top (like doing normal mode command `zt`) and sometime the current line position in the middle (like using normal mode command `zz`). That's kind of annoying, so I tried to use `winsaveview()` to make window view (current line position, etc) more consistent when exit Bufstop window.

I haven't tested it in split window, but I think it should be the same.

My `remove trailling whitespace` autocommand kicks in and that makes it looks like I changed a bunch of line, so please ignore that 😅

### CODE REVIEW

**Comment by mihaifm** [comment]
This is very promising if it works. I didn't know about `winsaveview`, guess there is something to learn every day. This issue bothers me too, but kind of learned to live with it. Let me test this, it would be a very useful fix.

**Comment by bruhtus** [comment]
> I didn't know about winsaveview, guess there is something to learn every day.

I didn't realize at first that I could initialize `winsaveview()` at the command until I saw [fzf.vim source code](https://github.com/junegunn/fzf.vim/blob/e34f6c129d39b90db44df1107c8b7dfacfd18946/plugin/fzf.vim#L65-L66) and I thought "maybe it can solve Bufstop current line position issue?" and then here we are.

> Let me test this, it would be a very useful fix.

Alright, let me know if there's some problem. I only apply `winsaveview()` in `Bufstop`, `BufstopPreview`, and `BufstopFast` because I thought it has something to do with Bufstop window, but if you think we should apply `winsaveview()` in every Bufstop command then I'll change the script again.

File: plugin/bufstop.vim [hunk 1]
```diff
@@ -179,9 +183,17 @@ function! s:BufstopWipeBuffer(bufnr)
   setlocal nomodifiable
 endfunction
 
+function! BufstopRestoreWinview()
+  q
+  if exists('b:bufstop_winview')
+    call winrestview(b:bufstop_winview)
+    unlet b:bufstop_winview
+  endif
+endfunction
```
Comment (mihaifm) [inline plugin/bufstop.vim:192]: You need to return to the previous window here (`wincmd p`), othewise your code won't work when two or more windows are present (split windows).

Comment (bruhtus) [reply plugin/bufstop.vim:192]: Ah I see, between `endif` and `endfunction` right? for single window (not using split window), there's no problem?

**Comment by bruhtus** [comment]
I think I found another issue, when deleting the current buffer (https://github.com/mihaifm/bufstop/commit/661288109d29465541614b9d020252d5bd3909ac), the window view changed back to how it was before (the current line position at the top like doing `zt` normal mode command).
I'm not sure where I should initialize the `winsaveview()` when delete the buffer, do you have any idea where should I initialize the `winsaveview()`?

Inside
```vim
if delkey
   call s:BufstopWipeBuffer(s:bufnr)
```
or inside
```vim
function! s:BufstopWipeBuffer(bufnr)
```
or maybe somewhere else?

**Comment by bruhtus** [comment]
> I think I found another issue, when deleting the first buffer ([6612881](https://github.com/mihaifm/bufstop/commit/661288109d29465541614b9d020252d5bd3909ac)), the window view changed back to how it was before (the current line position at the top like doing `zt` normal mode command).
> I'm not sure where I should initialize the `winsaveview()` when delete the buffer, do you have any idea where should I initialize the `winsaveview()`?
> 
> Inside
> 
> ```viml
> if delkey
>    call s:BufstopWipeBuffer(s:bufnr)
> ```
> 
> or inside
> 
> ```viml
> function! s:BufstopWipeBuffer(bufnr)
> ```
> 
> or maybe somewhere else?
> 
> EDIT:
> I think the dummy buffer causing the window view not restored when exiting Bufstop window.
> 
> Here's a step to reproduce that:
> 
> * open a file with vim
> * put the cursor at the bottom of the window by doing `L` normal mode command
> * open another file (using a buffer, not split window)
> * delete the recent opened file with Bufstop window
> * and then close Bufstop window using `BufstopDismissKey`
> 
> EDIT 2:
> When I delete the first buffer and it back to the second buffer, the second buffer doesn't have `b:bufstop_winview` variable even tho it is in opened.

I think I figure out where should I initialize `winsaveview()`, It's in `s:BufstopWipeBuffer(bufnr)` function, like this:
```vim
exe window . "wincmd w"
exe "silent b" candidate
if !exists('b:bufstop_winview')
  let b:bufstop_winview = winsaveview()
endif
```
CMIIW

File: plugin/bufstop.vim (hunk 1 above)
Comment (mihaifm) [reply plugin/bufstop.vim:192]: yes, this should work

**Comment by mihaifm** [comment]
> I think I figure out where should I initialize `winsaveview()`, It's in `s:BufstopWipeBuffer(bufnr)` function, like this:
> 
> ```viml
> exe window . "wincmd w"
> exe "silent b" candidate
> if !exists('b:bufstop_winview')
>   let b:bufstop_winview = winsaveview()
> endif
> ```

This looks good

**Comment by mihaifm** [comment]
Good staff, thanks for the contribution, it's an awesome fix. I will do some minor refactoring on top of these changes.

**Comment by bruhtus** [comment]
Glad I could help!

**Pull request merged**